### PR TITLE
As a  user using an echo server I would like to be able to send a message to the server. 

### PR DESCRIPTION
**Pull request by ephieo**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Echo Server
+# Echo echo.Server
```
Comment (Kotauror) [inline README.md:1]: Missing information on how to run and test it.

File: build.gradle [hunk 2]
```diff
@@ -13,9 +13,10 @@ repositories {
 dependencies {
```
Comment (Kotauror) [inline build.gradle:13]: It would be great if the build directory and the . dirs weren't merged into main.

File: src/main/java/echo/Client.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        runClient(args[0],Integer.parseInt(args[1]));
+    }
+
+    public static void runClient (String hostname, int portNumber){
+        try (
+                Socket clientSocket = new Socket(hostname, portNumber);
+                PrintWriter output = new PrintWriter(clientSocket.getOutputStream(), true);
+                BufferedReader input = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
+                BufferedReader stdInput = new BufferedReader(new InputStreamReader(System.in));
```
Comment (Kotauror) [inline src/main/java/echo/Client.java]: are they both needed?

File: src/main/java/echo/Client.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+            }
+
+        } catch (UnknownHostException e) {
+            System.err.println("Don't know about host " + hostname);
+            System.exit(1);
+        } catch (IOException e) {
+            System.err.println("Couldn't get I/O for the connection to " +
+                    hostname);
+            System.exit(1);
+        }
```
Comment (Kotauror) [inline src/main/java/echo/Client.java]: This is very good catching - explicit errors (different errors different messages), with a message and action (here - exit).

File: src/main/java/echo/Run.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+//package echo;
```
Comment (Kotauror) [inline src/main/java/echo/Run.java]: Is this file needed?

File: src/main/java/echo/Server.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+import java.net.*;
+
+
+public class Server {
+    public static void main(String[] args) throws IOException {
+        runEchoServer();
+    }
+
+    public static ServerSocket serverSocket = null;
+        public static Socket clientSocket = null;
```
Comment (Kotauror) [inline src/main/java/echo/Server.java]: I'm not good at Java, so excuse my simple question, but is this `null` assignment here needed? I understand that these are declared on top to avoid errors, but would it be enough to do: 

```
    public static ServerSocket serverSocket;
    public static Socket clientSocket;
```

File: src/main/java/echo/Server.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+
+
+    public static void runEchoServer() throws IOException {
+
+        try {
+            serverSocket = createServerSocket(7777);
```
Comment (Kotauror) [inline src/main/java/echo/Server.java]: Have you considered passing the port number to the server, to avoid hardcoding it here?

File: src/main/java/echo/Server.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+        print("Listening for connection...");
+
+        try {
+
+            clientSocket = connectClient(serverSocket);
+
+        } catch (IOException e) {
+            errorMessage("***Accept failed!!!***", e);
+
```
Comment (Kotauror) [inline src/main/java/echo/Server.java:43]: the indentation is very off in this file. Do you have a formatting tool for java?

File: src/main/java/echo/Server.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        } catch (IOException e) {
+            errorMessage("Can't listen on port number 7777", e);
+
+        }
+
+        print("Listening for connection...");
+
+        try {
+
```
Comment (Kotauror) [inline src/main/java/echo/Server.java]: I like the try - catch here too - it's very clear that there are several steps along the way that can fail, all with different messages. It would be clear to the user what went wrong, good job 👏

File: src/main/java/echo/Server.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        print("***Successful*** " + serverSocket);
+        return serverSocket;
+    }
+
+    public static Socket connectClient(ServerSocket serverSocket) throws IOException {
+        Socket clientSocket = serverSocket.accept();
+        return clientSocket;
+    }
+
+    public static void errorMessage(String message, IOException e) {
```
Comment (Kotauror) [inline src/main/java/echo/Server.java]: this was surprising to me - the `errorMessage` not only print it, but also exits the program. I think these two functionalities should be separate.

File: src/test/java/ServerTest.java [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    @Test
+    public void testingSocketSendsMessage() throws IOException {
+        Client client = new Client();
+        Server server = new Server();
+        System.out.println(server);
+
+        String fakeMessage = "Merry Christmas :)\n bye\n";
+//        String connectMessage = "nc localhost 7777\n";
```
Comment (Kotauror) [inline src/test/java/ServerTest.java]: remove commented out code (also below)

File: src/main/java/echo/Run.java (hunk 5 above)
Comment (ephieo) [reply src/main/java/echo/Run.java]: I was trying to create a file that ran the whole project I've changed it to `main.java`

File: src/main/java/echo/Server.java (hunk 7 above)
Comment (ephieo) [reply src/main/java/echo/Server.java]: Yes I can make that change.

File: src/main/java/echo/Server.java (hunk 8 above)
Comment (ephieo) [reply src/main/java/echo/Server.java:43]: Yes I have  a formatting tool i will format it and push it up.

File: README.md (hunk 1 above)
Comment (ephieo) [reply README.md:1]: done ✅

File: build.gradle (hunk 2 above)
Comment (ephieo) [reply build.gradle:13]: I tried to fix this and ended up losing some files, I'm not sure how to remove them from my repo. 
❓

File: src/main/java/echo/Client.java (hunk 3 above)
Comment (ephieo) [reply src/main/java/echo/Client.java]: Not too sure but I've removed the client now.

File: src/main/java/echo/Client.java (hunk 4 above)
Comment (ephieo) [reply src/main/java/echo/Client.java]: Thank you Justyna 👍

File: src/main/java/echo/Server.java (hunk 6 above)
Comment (ephieo) [reply src/main/java/echo/Server.java]: I don't think it was needed in that file but now that I've moved those variables to the `main.java` file they can't be passed down if they're not assigned a value. it gives me errors when I try. 
❓

**Review by Kotauror — APPROVED** [review]
